File: example/java/jimmer-sql2/src/main/java/org/babyfish/jimmer/sql2/example/bll/interceptor/input/BaseEntityDraftInterceptor.java
package org.babyfish.jimmer.sql2.example.bll.interceptor.input;

import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.babyfish.jimmer.sql2.example.model.common.BaseEntityDraft;
import org.babyfish.jimmer.sql2.example.model.common.BaseEntityProps;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;

/*
 * See JSqlClient.Builder.addDraftInterceptors()
 */
@Component
public class BaseEntityDraftInterceptor implements DraftInterceptor<BaseEntityDraft> {

    @Override
    public void beforeSave(BaseEntityDraft draft, boolean isNew) {
        if (!ImmutableObjects.isLoaded(draft, BaseEntityProps.MODIFIED_TIME)) {
            draft.setModifiedTime(LocalDateTime.now());
        }
        if (isNew && !ImmutableObjects.isLoaded(draft, BaseEntityProps.CREATED_TIME)) {
            draft.setCreatedTime(LocalDateTime.now());
        }
    }
}

File: example/kotlin/jimmer-sql-graphql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/graphql/bll/interceptor/input/BaseEntityDraftInterceptor.kt
package org.babyfish.jimmer.example.kt.graphql.bll.interceptor.input

import org.babyfish.jimmer.example.kt.graphql.entities.common.BaseEntity
import org.babyfish.jimmer.example.kt.graphql.entities.common.BaseEntityDraft
import org.babyfish.jimmer.kt.isLoaded
import org.babyfish.jimmer.sql.DraftInterceptor
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class BaseEntityDraftInterceptor : DraftInterceptor<BaseEntityDraft> {

    override fun beforeSave(draft: BaseEntityDraft, isNew: Boolean) {
        if (!isLoaded(draft, BaseEntity::modifiedTime)) {
            draft.modifiedTime = LocalDateTime.now()
        }
        if (isNew && !isLoaded(draft, BaseEntity::createdTime)) {
            draft.createdTime = LocalDateTime.now()
        }
    }
}